web/modules/custom/nidirect_webforms/src/Controller/PrisonVisitBookingJsonApiController.php

```diff
@@ -0,0 +1,216 @@
+   */
+  private function isValidRequest() {
+
+    // API restricted to specific client IP addresses and the client
+    // must send secret token.
+    $client_ip = $this->request->getClientIp();
+    $client_token = $this->request->headers->get('X-Auth-Token');
+
+    // Allowed client IPs.
+    $allowed_ip_addresses = explode(',', getenv('PRISON_VISITS_API_PERMITTED_IPS'));
+    $allowed_ip_addresses = array_map('trim', $allowed_ip_addresses);
+
+    // Allowed tokens.
+    $allowed_tokens = explode(',', getenv('PRISON_VISITS_API_PERMITTED_TOKENS'));
+    $allowed_tokens = array_map('trim', $allowed_tokens);
+
+    if (!$allowed_ip_addresses || !$allowed_tokens) {
+      $this->getLogger('prison_visits')->warning('One or more environment variables are missing: PRISON_VISITS_API_PERMITTED_IPS, PRISON_VISITS_API_PERMITTED_TOKENS');
+    }
+
+    // Is ip and token allowed?
+    $ip_is_allowed = !empty($client_ip) && in_array($client_ip, $allowed_ip_addresses);
+    $token_is_allowed = !empty($client_token) && in_array($client_token, $allowed_tokens);
+
+    // Return a session if the request passes the validation.
+    if ($ip_is_allowed && $token_is_allowed) {
+      return TRUE;
+    }
+
+    return FALSE;
+  }
+
+  /**
+   * Helper function to cache posted json data.
+   *
+   * Data is cached and returned only if it is within size limits can
+   * and can be json decoded.
+   */
+  private function cacheJsonData() {
+
+    $data = [];
+    $content = $this->request->getContent();
+
+    // Return no data if posted json content exceeds 100,000 bytes.
+    if (strlen($content) > 100000) {
+      $this->getLogger('prison_visits')->error('prison_visit_slots_data exceeds 100000 bytes');
+      return [];
+    }
+    elseif (!empty($content)) {
+
+      $now = new \DateTime('now');
+      $expire = $now->modify('+1 week');
+
+      // If content can be json decoded, write the decoded data to
+      // cache and the json content to a file.
+      if ($data = json_decode($content, TRUE)) {
+
+        // Write data to cache.
+        $this->cache->set('prison_visit_slots_data', $data, $expire->getTimestamp());
+        $this->getLogger('prison_visits')->info('prison_visit_slots_data cache updated.');
+
+        // Write content to file.
+        /** @var \Drupal\file\FileRepositoryInterface $fileRepository */
+        $fileRepository = \Drupal::service('file.repository');
+        $directory = 'private://nidirect_webforms';
+        $filepath = $directory . '/prison_visit_slots_data.json';
+
+        // Create file if it doesn't exist.
+        $file = $fileRepository->loadByUri($filepath);
+        if (empty($file)) {
+          /** @var \Drupal\Core\File\FileSystemInterface $file_system */
+          $file_system = \Drupal::service('file_system');
+          $file_system->prepareDirectory($directory, FileSystemInterface::CREATE_DIRECTORY | FileSystemInterface::MODIFY_PERMISSIONS);
+          $file = File::create([
```

Comment:
This will create a file entity entry, do we want this or do we just want to write the file to the filesystem?

Reply:
I think there is a setting that causes Drupal to delete unused files on cron.  Whilst we don't use that setting, I was thinking it might be safer to make this a managed file.